File: week2/MarshalLee/MinStack.cs
using System;

public class MinStack
{
    public class MinStack
    {

        private readonly Stack<KeyValuePair<int, int>> Stack = new();

        public MinStack()
        {
        }

        public void Push(int val)
        {
            int min = Stack.Count == 0 ? val : Math.Min(Stack.Peek().Value, val);

            Stack.Push(new KeyValuePair<int, int>(val, min));
        }

        public void Pop()
        {
            Stack.Pop();
        }

        public int Top()
        {
            return Stack.Peek().Key;
        }

        public int GetMin()
        {
            return Stack.Peek().Value;
        }
    }
}

File: week2/kuzey/MinStack.java
class Node {
    int data;
    int min;
    Node next;

    Node(int data, int min, Node next) {
        this.data = data;
        this.min = min;
        this.next = next;
    }
}

class MinStack {
    Node head;
    int size;

    public MinStack() {
        head = null;
        size = 0;
    }

    public void push(int val) {
        if (head == null) {
            head = new Node(val, val, null);
        } else {
            if (val < getMin()) {
                head = new Node(val, val, head);
            } else {
                head = new Node(val, getMin(), head);
            }
        }
    }

    public void pop() {
        head = head.next;
    }

    public int top() {
        return head.data;
    }

    public int getMin() {
        return head.min;
    }
}
